### PR TITLE
Fixed issue with mesh vertex buffers not updating properly

### LINKED ISSUE #6483: Meshes are not created in the Mesh component
**Meshes are not created in the Mesh component**

When dynamically creating/deleting meshes, there comes a point (right after the start) when the meshes stop being created without any errors. They simply do not exist.
If you create/delete meshes over a period of time, a crash occurs in the engine's CPP-code.

**To Reproduce**
1. Start example project.
2. Look around - rotate the camera with a mouse touch. Adjust your view so that you can see more of the landscape.
3. Move forward by pressing the arrows on the screen or using WASD.
4. See error - empty holes instead of landscape chunks.

image

**Expected behavior**
This correct screenshot is from the first launch, when there was no mesh management yet.
This is what it should look like:

image

**Defold version:**
  - 1.2.192

**Platforms:**
 - Platforms: Android, Windows 10, not tested on the others

**Minimal repro case project:**
[example_project.zip](https://github.com/defold/defold/files/8311930/example_project.zip)

**Logs (OPTIONAL):**
[log.zip](https://github.com/defold/defold/files/8311909/log.zip)

### PR DESCRIPTION
**Pull request by JCash**
Under certain circumstances meshes did not update correctly even though the buffers changed. This fix ensures that a mesh is updated when the buffer content changes.

### CODE REVIEW

File: engine/gamesys/src/dmsdk/gamesys/resources/res_buffer.h [hunk 1]
```diff
@@ -28,7 +28,8 @@ namespace dmGameSystem
         dmBufferDDF::BufferDesc* m_BufferDDF;
         dmBuffer::HBuffer        m_Buffer;
         dmhash_t                 m_NameHash;
-        uint32_t                 m_ElementCount;
+        uint32_t                 m_ElementCount;    // The number of vertices
+        uint32_t                 m_Stride;          // The vertex size (bytes)
```
Comment (JCash) [inline engine/gamesys/src/dmsdk/gamesys/resources/res_buffer.h:32]: Better to keep it in the buffer resource and not the mesh resource

File: engine/gamesys/src/dmsdk/gamesys/resources/res_mesh.h [hunk 2]
```diff
@@ -34,8 +34,6 @@ namespace dmGameSystem
 
         dmGraphics::HVertexDeclaration  m_VertexDeclaration;
         dmGraphics::HVertexBuffer       m_VertexBuffer;
-        uint32_t                        m_ElementCount;
-        uint32_t                        m_VertSize;
```
Comment (JCash) [inline engine/gamesys/src/dmsdk/gamesys/resources/res_mesh.h:38]: Can be retrieved from the buffer resource

File: engine/gamesys/src/gamesys/components/comp_mesh.cpp [hunk 3]
```diff
@@ -158,6 +157,19 @@ namespace dmGameSystem
...
+        HashState32 state;
+        dmHashInit32(&state, false);
+
+        uint32_t version;
+        dmBuffer::GetContentVersion(br->m_Buffer, &version);
+        dmHashUpdateBuffer32(&state, &br->m_Buffer, sizeof(br->m_Buffer)); // The handle is not a pointer, but a version+index
+        dmHashUpdateBuffer32(&state, &version, sizeof(version));
+
+        return dmHashFinal32(&state);
+    }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_mesh.cpp:171]: One of the issues was that the buffer content version starts at 0 and counts upwards, and many other buffers may have a similar version.

So detecting a changed buffer requires more than the content version.

File: engine/gamesys/src/gamesys/components/comp_mesh.cpp [hunk 4]
```diff
@@ -285,8 +289,8 @@ namespace dmGameSystem
             dmHashUpdateBuffer32(&state, &texture, sizeof(texture));
         }
 
-        dmGameSystem::BufferResource* br = GetVerticesBuffer(component, component->m_Resource);
-        dmHashUpdateBuffer32(&state, &br->m_Version, sizeof(br->m_Version));
+        BufferResource* br = GetVerticesBuffer(component, resource);
+        dmHashUpdateBuffer32(&state, &br->m_NameHash, sizeof(br->m_NameHash));
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_mesh.cpp:293]: The hash is for batching, and as such we want to batch on the resource itself, not the content version which might be the same between the instances using different buffers.

File: engine/gamesys/src/gamesys/components/comp_mesh.cpp [hunk 5]
```diff
@@ -425,11 +432,26 @@ namespace dmGameSystem
...
+                assert(info != 0);
 
-            if (component.m_ReHash || (component.m_RenderConstants && dmGameSystem::AreRenderConstantsUpdated(component.m_RenderConstants)))
+                if (info->m_Version != component.m_BufferVersion)
+                {
+                    info->m_Version = component.m_BufferVersion;
+
+                    CopyBufferToVertexBuffer(br->m_Buffer, info->m_VertexBuffer, br->m_Stride, br->m_ElementCount, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
+                }
+            }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_mesh.cpp:449]: Moved this code to the Update(), from the RenderUpdate().

File: engine/gamesys/src/gamesys/resources/res_mesh.cpp [hunk 6]
```diff
@@ -122,9 +122,7 @@ namespace dmGameSystem
         }
     }
 
-    bool BuildVertexDeclaration(BufferResource* buffer_resource,
-        dmGraphics::HVertexDeclaration* out_vert_decl,
-        uint32_t* out_elem_count, uint32_t* out_vert_size)
+    bool BuildVertexDeclaration(BufferResource* buffer_resource, dmGraphics::HVertexDeclaration* out_vert_decl)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_mesh.cpp:125]: No need for the function to handle variables that can be retrieved from the buffer resource.

File: engine/gamesys/src/gamesys/scripts/script_buffer.cpp [hunk 7]
```diff
@@ -749,7 +749,7 @@ namespace dmGameSystem
         *s = 0;
         uint32_t version = 0;
         dmBuffer::GetContentVersion(hbuffer, &version);
-        dmSnPrintf(buf, sizeof(buf), "buffer.%s(count = %d, version = %u", SCRIPT_TYPE_NAME_BUFFER, out_element_count, version);
+        dmSnPrintf(buf, sizeof(buf), "buffer.%s(count = %d, version = %u, ", SCRIPT_TYPE_NAME_BUFFER, out_element_count, version);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_buffer.cpp:752]: Readability fix

File: engine/gamesys/src/gamesys/components/comp_mesh.cpp [hunk 8]
```diff
@@ -168,17 +180,17 @@ namespace dmGameSystem
...
     {
         dmGraphics::HVertexBuffer vertex_buffer = GetVertexBuffer(world, br->m_NameHash);
         if (!vertex_buffer)
         {
             vertex_buffer = AllocVertexBuffer(world, world->m_GraphicsContext);
-            AddVertexBufferInfo(world, br->m_NameHash, vertex_buffer, br->m_Version); // ref count == 1
+            AddVertexBufferInfo(world, br->m_NameHash, vertex_buffer, version); // ref count == 1
 
-            uint32_t elem_count = br->m_ElementCount;
+            VertexBufferInfo* info = world->m_ResourceToVertexBuffer.Get(br->m_NameHash);
```
Comment (britzl) [inline engine/gamesys/src/gamesys/components/comp_mesh.cpp:191]: Unused?

**Review by britzl — APPROVED** [review]

**Pull request merged**